src/erleans_timer.erl

```diff
@@ -0,0 +1,200 @@
+        _ ->
+            {error, no_timer_to_cancel}
+    end.
+
+%%% in order to be able to recover the list of timers when the grain
+%%% starts to go down and then is reactivated before all of the
+%%% in-process timer ticks can complete, we need to have a slightly
+%%% different form of cancel, which leaves the timer information
+%%% structure intact so we can restart them if we need to.
+recoverable_cancel() ->
+    case get(?key) of
+        undefined ->
+            {error, no_timer_to_cancel};
+        Map when is_map(Map) ->
+            maps:fold(fun(Pid, _V, Acc) ->
+                              Pid ! ?cancel,
+                              Acc
+                      end, beep, Map),
+            erase(?key),
+            put(?recovery, Map)
+    end.
+
+check() ->
+    case get(?recovery) of
+        undefined ->
+            finished;
+        Map when is_map(Map) ->
+            AnythingAlive =
+                maps:fold(fun(_Pid, _V, true) ->
+                                  true;
+                             (Pid, _V, _) ->
+                                  is_process_alive(Pid)
+                          end, false, Map),
+            case AnythingAlive of
+                false ->
+                    finished;
+                _ ->
+                    pending
+            end
+    end.
+
+recover() ->
+    case get(?recovery) of
+        undefined ->
+            ok;
+        Map when is_map(Map) ->
+            NewMap = maps:fold(fun(_Pid, Timer, Acc) ->
+                                       NewPid = start_timer(Timer#timer.period, Timer),
+                                       Acc#{NewPid => Timer}
+                               end, #{}, Map),
+            erase(?recovery),
+            put(?key, NewMap)
+    end.
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% internal functions %%%
+%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+loop(FireTime, #timer{grain = Grain,
+                      callback = Callback,
+                      args = Args,
+                      period = Period} = Timer) ->
+    receive
+        ?cancel ->
+            unlink(Grain),
+            ok;
+        ?tick ->
+            try Callback(Grain, Args) of
+                _ ->
+                    case Period of
+                        never ->
+                            unlink(Grain),
+                            ok;
+                        _ ->
+                            NextFire = FireTime + Period,
+                            erlang:send_after(NextFire, self(), ?tick,
+                                              [{abs, true}]),
+                            loop(NextFire, Timer)
+                    end
+            catch Class:Error ->
+                    Grain ! {erleans_timer_error, Class, Error},
+                    unlink(Grain)
+            end;
+        Msg ->
+            Grain ! {erleans_timer_unexpected_msg, Msg}
+    end.
+
+start_timer(StartTime, #timer{grain = Grain} = Timer) ->
+    spawn(fun() ->
+                  link(Grain),
```

Comment:
yeah. it was originally a `monitor`, then I didn't fold them together when I changed it.